/* @brief Implements serialization for measurements on the Soil Power Sensor
 * 
 * Each individual measurement has all identifying information for someone to
 * determine when the measurement happened, who performed the measurement, what
 * was being measured, and the measurement itself. In the scheme of the Soil
 * Power Sensor, single measurements are taken, then saved to NVRAM for a later
 * upload. This was perferred over using "repeated" types as the data would have
 * to be uploaded in blocks of LoRaWAN due to payload size limitations.
 * 
 * @author John Madden <jmadden173@pm.me>
 * @date 2023-11-21
 */

syntax = "proto3";

import "timestamp.proto";

/* Data shared between all measurement messages */
message MeasurementMetadata {
    // id of the cell measured
    uint32 cell_id = 1;
    // id of the logging device
    uint32 logger_id = 2;
    // timestamp of the measurement
    google.protobuf.Timestamp ts = 3;
}

/* Power measurement message. Voltage and current can be digitially combined to
 * obtain power. */
message PowerMeasurement {
    MeasurementMetadata meta = 1;

    // voltage
    double voltage = 2;
    // current
    double current = 3;
}

/* Teros12 measurement message */
message Teros12Measurement {
    MeasurementMetadata meta = 1;

    // raw volumetric water content
    float vwc_raw = 2;
    // calibrated volumetric water content
    float vwc_adj = 3;
    // temperature in celcious
    float temp = 4;
    // electrical conductivity
    uint32 ec = 5;
}